lib/jxl/enc_ans.cc

```diff
@@ -1496,14 +1496,33 @@ size_t BuildAndEncodeHistograms(const HistogramParams& params,
     uint_config = HybridUintConfig(10, 0, 0);
   }
   for (size_t i = 0; i < tokens.size(); ++i) {
-    for (size_t j = 0; j < tokens[i].size(); ++j) {
-      const Token token = tokens[i][j];
-      total_tokens++;
-      uint32_t tok, nbits, bits;
-      (token.is_lz77_length ? codes->lz77.length_uint_config : uint_config)
-          .Encode(token.value, &tok, &nbits, &bits);
-      tok += token.is_lz77_length ? codes->lz77.min_symbol : 0;
-      builder.VisitSymbol(tok, token.context);
+    if (codes->lz77.enabled) {
+      for (size_t j = 0; j < tokens[i].size(); ++j) {
+        const Token& token = tokens[i][j];
+        total_tokens++;
+        uint32_t tok, nbits, bits;
+        (token.is_lz77_length ? codes->lz77.length_uint_config : uint_config)
+            .Encode(token.value, &tok, &nbits, &bits);
+        tok += token.is_lz77_length ? codes->lz77.min_symbol : 0;
+        builder.VisitSymbol(tok, token.context);
+      }
+    } else if (num_contexts == 1) {
+      Histogram& singlehisto = const_cast<Histogram&>(builder.Histo(0));
```

Comment:
This branch feels like it could be quite dangerous... Does it give significant benefit?

Reply:
Without that branch:
4064 x 2704, geomean: 32.15 MP/s [24.28, 34.25], 30 reps, 4 threads.

With that branch:
4064 x 2704, geomean: 35.27 MP/s [26.83, 36.66], 30 reps, 4 threads.

The thing with -e 1 is that it is spending so little overall effort that even silly things like this can make a noticeable difference.

This is what perf top looks like after this pull request:
```
  33.65%  cjxl                                [.] jxl::WriteTokens
  17.95%  cjxl                                [.] jxl::EncodeModularChannelMAANS
   9.28%  cjxl                                [.] jxl::BuildAndEncodeHistograms
   8.61%  libc-2.32.so                        [.] __memmove_avx_unaligned_erms
   6.58%  cjxl                                [.] jxl::(anonymous namespace)::float_to_int
   4.50%  cjxl                                [.] jxl::ThreadPool::RunCallState<jxl::Status (unsigned long), jxl::FwdRCT(jxl::Image&, unsigned long, unsigned long, jxl::ThreadPool*)::$_
   3.55%  cjxl                                [.] std::vector<jxl::Token, std::allocator<jxl::Token> >::_M_fill_insert
   1.71%  [kernel]                            [k] clear_page_rep
   1.16%  [kernel]                            [k] do_user_addr_fault
   0.68%  cjxl                                [.] jxl::BitWriter::Write
```

Looks like doing resize() on the token vector is still not that cheap, given that it just has to zero-initialize the memory (but it's using more generic fill code instead). The float to int and memmoves should mostly be avoidable. The top 3 functions are of course needed, but any small speedup in there will have a noticeable effect on the overall encode time.

Reply:
Makes sense. Can we modify the histogram builder interface so that it's clear that something like this could happen?

Reply:
Maybe do that in a later pull request? There might be an opportunity to speed up the histogram building a bit more by parallelizing it: basically split the token list in N sublists (one per thread), compute N histograms, then add them together (there's already a function for that, I've seen). This is one of the steps that are currently not parallelized at all, while most of the histogram building happens globally so it would make sense to parallelize it.

Reply:
I'd rather also move adding this branch to a later PR then - I have strong feelings against breaking an interface's contract this way ;)

Reply:
Would it be OK to just change `HistogramBuilder.Histo()` to not return a `const` Histogram?

Reply:
yes if it goes together with a nice comment :)

Reply:
Did it in a different way, just using ` builder.VisitSymbol(tok, 0)` instead of `builder.VisitSymbol(tok, token.context)` and letting the compiler do a specialized inlining of that. That's cleaner than the manual specialization with the nasty `const_cast`.